### PR TITLE
[Feature] Support for Default Model Id

### PR DESCRIPTION
**Pull request by vibrantvarun**
### Description
This PR provides a functionality to client to skip passing model id in the neural search query request. 

### Issues Resolved
[70](https://github.com/opensearch-project/neural-search/issues/70)

### Check List
- [x] New functionality includes testing.
    - [x] All tests pass
- [x] New functionality has been documented.
    - [x] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by navneet1v** [comment]
@vibrantvarun remove the skip changelog tag and put the info in change log.

**Comment by navneet1v** [comment]
@vibrantvarun fill the checklist too

**Comment by navneet1v** [comment]
@vibrantvarun add documentation on all public classes and public functions

File: src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java
Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java:20]: we can use @AllArgsConstructor here

File: src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java
Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java:26]: as per my understanding the logic here should be :
1. If model id is provided in query then do nothing.
2. If model id is present in the field map, use that model id
3. If no field is not in the field map, then use the model id provided in the SearchRequestProcessor. 

The current code doesn't seem to work like that. Can you share some light on the logic.

File: src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java
Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java:19]: do we really need this class cannot we directly get it from any OpenSearch class?

File: src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java
Comment (navneet1v) [reply src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java:26]: cool, lets fix the logic.

File: src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java:19]: I would suggest let's keep this class as it provides abstraction from underlying opensearch cluster, also if i have to remove this then I have initialize cluster service variable in the create components and then have take it all the way to neural query builder.

Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java:19]: Because Version data comes from clusterService.state().nodes().getMinNodeVersion()

File: src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java
Comment (navneet1v) [inline src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java:98]: lets move this function after all public functions.

File: src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java
Comment (navneet1v) [inline src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java:89]: do we really need this? what are really testing with these, as these are integration tests

File: src/test/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtilTests.java
Comment (navneet1v) [inline src/test/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtilTests.java:18]: try to put your tests function names like

test<WhatYourTesting>_when<ABCCondition>_then<XYZResult>

File: src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorIT.java
Comment (navneet1v) [inline src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorIT.java:29]: why public?

File: src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorIT.java
Comment (navneet1v) [inline src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorIT.java:50]: use @SneakyThrows

File: src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java
Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java:89]: Yes we need this because when we get node version from cluster. this.clusterService.getState.getNode.getMinNodeVersion.

While executing integ test case of NeuralQueryProcessor, **this.clusterService** throws null pointer if we do not mock it.

**Review by navneet1v — COMMENTED** [review]
Can we add Unit tests for the visitor class. and test all the scenarios

File: src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java
Comment (navneet1v) [reply src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java:89]: But these are integration steps, why do I need to mock it?

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:83]: Can we specify the exception as IllegalArgumentException instead of Exception?

File: src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java:126]: Can this be simplified to:
```
if (modelId != null) {

}
```

File: src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java:276]: One problem with this is that the function name does not indicate that it has anything to do with default model ID. I would recommend parametrizing this with the version.

File: src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java:18]: nit: "It visits the each and every component of query buikder tree." -> "It visits each and every component of query builder tree."

File: src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java
Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/common/BaseNeuralSearchIT.java:89]: Oops I used the wrong word here, instead of mock I should I have used initialize. Opensearch has created a test suite for integ test called OpenSearchIntegTestCase. We use is BaseNeuralSearchIT in our test cases. Now in OpenSearchIntegTestCase they [initialize](https://github.com/opensearch-project/OpenSearch/blob/8f4f9951de817a58d18815c5dd35896dae2bced1/test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java#L654) clusterService. In our case we don't initalize so that is the reason I was getting null in NeuralSearchClusterUtil. Now I have initialized.

File: src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java:126]: Yes, I was planning to do the same. But @navneet1v earlier suggested to put this clusterOnorAfter checks so to maintain the similarity in the work I used this one. Any thoughts @navneet1v ?

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:19]: Can we have more descriptive text here, this one looks more like a class name

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:26]: is this the agreed name for this processor? It sounds like a processor suppose to execute the query, maybe something describing actual functionality like enriching query with missing data.

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:66]: this should be a separate class, please check existing factory classes for text_embedding for example

File: src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java:24]: if those are only set in constructor please make them final

File: src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java:99]: please add a comment on what the new logic is. You may move it to separate private method `getModelId` and put it in the method's description

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:28]: private?

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:30]: private? Also, could we add a comment for this member?

Also, why is this map <String, Object> and not <String, String>

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:42]: Can this just be package-private or private? Will this class be sub-classed?

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:84]: Is this cast safe?

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:47]: Could we add @Nullable annotation to indicate these can be null?

File: src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java
Comment (jmazanec15) [inline src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java:54]: This seems dangerous to suppress all exceptions. This looks like it was taken from https://github.com/opensearch-project/k-NN/blob/ca5e483e1e70abdc19196e1018b3a7fd06908bf6/src/main/java/org/opensearch/knn/index/KNNClusterUtil.java#L46-L56. @martin-gaievski What was reasoning behind doing this?

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:26]: I personally think this current one looks good, we are anyway explaining everything in docs and other places. Type can  be a little lighter as well

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:66]: Text embedding processor is a ingest processor. Neural Query Processor is a search request processor. I have followed the same template of search request processor how they are implemented in OpenSearch. We should follow the same template of OS

[FilterQueryRequestProcessor](https://github.com/opensearch-project/OpenSearch/blob/main/modules/search-pipeline-common/src/main/java/org/opensearch/search/pipeline/common/FilterQueryRequestProcessor.java)

File: src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java:99]: Yes I will add the comment. Moving into method looks unecessory.

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:30]: Because [ConfigurationUtils](https://github.com/opensearch-project/OpenSearch/blob/d656e3db592f29466d35452867caa241f5429485/server/src/main/java/org/opensearch/ingest/ConfigurationUtils.java#L359) The class from Opensearch which is used to retrieve map from processor is generic and it Map<String, Object> defined in it. ConfigurationUtils is the designed to get values from the processor in OS. So I am using the same class

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:84]: Yeah I can put a null pointer check

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:26]: I would go with something like `enriching_query_defaults`, that is still light to me. The docs will be there for sure, what I mean is that having just `neural_query` gives wrong impression that we're executing that query, it's not about less details but about wrong direction. Plus with more generic name we can re-use that same processor for defaults in other fields.

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:66]: I don't see a reason why to follow that pattern, what makes more sense to me for decision is if that class has important and complex logic/responsibility, or not. We've done a separate class for [normalization processor factory](https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/processor/factory/NormalizationProcessorFactory.java), processor is a search processor. 
If you think factory for your processor is simple we can leave it as inner class

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:26]: makes sense.

File: src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java:54]: if I'm not mistaken, the idea was to allow cluster work with a `current` version instead of failing, maybe that core api was throwing some exception at that time. I checked core code, seems it's pretty straightforward now, I think we can get rid of that catch_all approach

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:66]: Yes it is simple.

File: src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/query/visitor/NeuralSearchQueryVisitor.java:24]: I think we should not keep it final because either of the value can be null. and variables with final keyword cannot have a value null

File: src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/util/NeuralSearchClusterUtil.java]: Taken care off now

**Comment by navneet1v** [comment]
@vibrantvarun can we fix the github action: codecov/patch and codecov/project 

seems like the coverage has dropped

File: src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorTests.java
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorTests.java:19]: please use naming convention from other tests - whatYouTesting_conditions_expectation, I think <conditions> may be optional sometimes, when it's obvious, but other two are must

File: src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorTests.java
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorTests.java:25]: why do we need this?

File: src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorTests.java
Comment (martin-gaievski) [inline src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorTests.java:39]: nit: you can use @sneakythrow for exceptions in tests

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java:28]: I would rename class to something closer to a processor type name, although I left the final call to you

**Review by martin-gaievski — COMMENTED** [review]
few minor things, over looks good

File: src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorTests.java
Comment (vibrantvarun) [reply src/test/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessorTests.java]: To check exception

File: src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java
Comment (navneet1v) [reply src/main/java/org/opensearch/neuralsearch/processor/NeuralQueryProcessor.java]: @martin-gaievski Enriching Query Processor is too generic. Can we rename it to NeuralQueryEnrichingProcessor. In my understanding we should limit the scope of this processor to only neural queries. Nothing else. If we go too generic then this will become a default processor to put any kind of enriching which will bloat up the responsibilities of this processor.

File: src/main/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessor.java
Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessor.java:33]: neural_query_enricher name seems better here

File: src/main/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessor.java
Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessor.java:93]: why config.remove?

Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessor.java:93]: There is a better way to get the string. Check this: https://github.com/opensearch-project/neural-search/blob/main/src/main/java/org/opensearch/neuralsearch/processor/factory/TextEmbeddingProcessorFactory.java#L40C9-L40C89

File: src/main/java/org/opensearch/neuralsearch/plugin/NeuralSearch.java
Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/plugin/NeuralSearch.java:83]: We should start getting rid of these kind of inits and move towards singleton pattern without this kind of inits. 

May be an AI for maintainers

File: src/test/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessorIT.java
Comment (navneet1v) [inline src/test/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessorIT.java:47]: we should delete the index too.

File: src/test/resources/processor/IndexMappings.json
Comment (navneet1v) [inline src/test/resources/processor/IndexMappings.json:76]: why we need these new fields? why we cannot use older fields?

Comment (vibrantvarun) [reply src/test/resources/processor/IndexMappings.json:76]: Because they don't have text matching

File: src/main/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessor.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessor.java:33]: @martin-gaievski I am finally going with neural_query_enricher

File: src/main/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessor.java
Comment (vibrantvarun) [reply src/main/java/org/opensearch/neuralsearch/processor/EnrichingQueryDefaultProcessor.java:93]: Took reference from OS core

**Review by martin-gaievski — APPROVED** [review]

File: src/test/resources/processor/IndexMappings.json
Comment (navneet1v) [reply src/test/resources/processor/IndexMappings.json:76]: do we need text matching?

Comment (vibrantvarun) [reply src/test/resources/processor/IndexMappings.json:76]: yeah i wrote a test case where I am passing querytext as hello world. If I don't add this field here then it will fail

**Review by navneet1v — APPROVED** [review]

File: src/test/resources/processor/IndexMappings.json
Comment (martin-gaievski) [inline src/test/resources/processor/IndexMappings.json:73]: this need to have a `lucene` as engine, otherwise it uses default which is `nmslib` and that fails, as nmslib is not part of the knn zip fopr integ tests @vibrantvarun

Comment (heemin32) [reply src/test/resources/processor/IndexMappings.json:73]: Where is this field used by the way?

**Pull request merged**
